Python Transaction Processor

sofiter opened the pull request:
Missing transfer action (for now)

sofiter committed:
python processor

Python processor for the simplewallet (not complete)

sofiter committed:
Update py_proc.py

sofiter committed:
Update py_proc.py

sofiter committed:
Add files via upload

sofiter committed:
Update py_proc.py

sofiter committed:
Delete sw_processor.py

askmish commented:
Thanks for a new python transaction processor.

To help maintain logs, could you:
1. squash all commits into one single commit with a commit message(like "Add Python Transaction Processor") and a commit body mentioning the features added in the commit(like "A simplewallet python transaction processor with deposit and withdraw features").

2.  add a [dco ](http://developercertificate.org/) sign-off(Signed-off-by) with your github email id, in the commit message.

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+sw_namespace = _hash(FAMILY_NAME.encode('utf-8'))[0:6]
+
+
+class SWTransactionHandler(TransactionHandler):
```
Lets use the "SimpleWalletTransactionHandler" here

sofiter replied:
Done

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+
+        if operation == "transfer":
+            self._make_transfer(context, amount, to_key, from_key)
+
```
What if the operation doesn't match any of the above conditions. Perhaps we'll need to handle that here.

sofiter replied:
I changed it to if..elif..else statement

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+            self._make_transfer(context, amount, to_key, from_key)
+
+    def _make_deposit(self, context, amount, from_key):
+
```
could remove this empty line

sofiter replied:
Of course, removed all the unnecessary empty lines

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+        LOGGER.info('Got the key {} and the wallet key {} '.format(from_key, wallet_key))
+
+        current_entry = context.get_state(wallet_key)
+        balance = str(current_entry)
```
You really don't need to convert 'current_entry' to 'str'
get_state will contain a list if the address/wallet_key exists or be empty if doesn't exist yet.

sofiter replied:
Done

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+        wallet_key = self._get_wallet_key(from_key)
+        LOGGER.info('Got the key {} and the wallet key {} '.format(from_key, wallet_key))
+
+        current_entry = context.get_state(wallet_key)
```
get_state needs a list of addresses.
The syntax should be: `context.get_state([wallet_key])`

askmish replied:
Here's a reference: https://github.com/hyperledger/sawtooth-core/blob/v1.0.4/sdk/python/sawtooth_sdk/processor/context.py

sofiter replied:
Thank you, it really helped to review the context

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+            LOGGER.info('No previous deposits, creating new deposit {} '.format(from_key))
+            new_balance = amount
+        else:
+            new_balance = amount + balance
```
We shouldn't be adding a string to a number.

sofiter replied:
No comments :-)

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,138 @@
+        state_data = new_balance.encode()
+
+        addresses = context.set_state(
+            {self._get_wallet_key(from_key): state_data})
```
No need to generate the key again. Just reuse the wallet_key you got in line 72 of this file.

sofiter replied:
Right!

sofiter committed:
Add Python Transaction Processor

A simplewallet python transaction processor with deposit and withdraw features
Signed-off-by: Sophia Terzi <sophia.terzi@gmail.com>

sofiter commented:
Thank you for the review and your suggestions. I believe I fixed everything you mentioned, please take a look and inform me if I missed anything. I will be back soon with the transfer operation to complete the processor

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,126 @@
+        self._namespace_prefix = namespace_prefix
+
+    @property
+    def family_name(self):
```
Lets use the `FAMILY_NAME` you've defined in line 24, here instead of using a new string.

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,126 @@
+    def _make_deposit(self, context, amount, from_key):
+        wallet_key = self._get_wallet_key(from_key)
+        LOGGER.info('Got the key {} and the wallet key {} '.format(from_key, wallet_key))
+        current_entry = context.get_state([wallet_key])
```
again, get_state() returns a list. You'll have to handle it accordingly.

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,126 @@
+    def _make_withdraw(self, context, amount, from_key):
+        wallet_key = self._get_wallet_key(from_key)
+        LOGGER.info('Got the key {} and the wallet key {} '.format(from_key, wallet_key))
+        current_entry = context.get_state([wallet_key])
```
get_state() returns a list

sofiter committed:
Update py_proc.py

sofiter committed:
Add Python Transaction Processor

Added transfer functionality

sofiter committed:
Add python transaction processor

A simplewallet python transaction processor with deposit, withdraw and transfer features
Signed-off-by: Sophia Terzi <sophia.terzi@gmail.com>

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,145 @@
+        if len(addresses) < 1:
+            raise InternalError("State Error")
+
+        def _make_transfer(self, context, transfer_amount, to_key, from_key):
```
Indentation alignment. Could you use the run_lint for sawtooth-core here. It would help you verify the code lints.

askmish commented on py_proc.py:
```diff
@@ -0,0 +1,145 @@
+        return _hash(sw_namespace.encode('utf-8'))[0:6] + _hash(from_key.encode('utf-8'))[0:64]
+
+def main():
+    processor = TransactionProcessor(url='tcp://localhost:4004')
```
Could you change this to "validator:4004" I will be creating dockerfiles for this python TP.

askmish replied:
Also could you do rebase and squash all your commits into one or two logical commits.

sofiter committed:
Final - Python Transaction Processor

A simplewallet python transaction processor with deposit, withdraw and transfer features
Signed-off-by: Sofia Terzi <sophia.terzi@gmail.com>
